src/accountsdb/download.zig

```diff
@@ -0,0 +1,652 @@
+    my_shred_version: usize,
+    my_pubkey: Pubkey,
+    blacklist: []Pubkey,
+    valid_peers: *std.ArrayList(PeerSnapshotHash),
+) PeerSearchResult {
+    valid_peers.clearRetainingCapacity();
+
+    var result = PeerSearchResult{};
+    search_loop: for (contact_infos) |*peer_contact_info| {
+        const is_me = peer_contact_info.pubkey.equals(&my_pubkey);
+        if (is_me) {
+            result.is_me_count += 1;
+            continue;
+        }
+
+        const matching_shred_version = my_shred_version == peer_contact_info.shred_version or my_shred_version == 0;
+        if (!matching_shred_version) {
+            result.invalid_shred_version += 1;
+            continue;
+        }
+        _ = peer_contact_info.getSocket(socket_tag.RPC) orelse {
+            result.no_rpc_count += 1;
+            continue;
+        };
+        const gossip_data = table.get(.{ .SnapshotHashes = peer_contact_info.pubkey }) orelse {
+            result.no_snapshot_hashes_count += 1;
+            continue;
+        };
+        const snapshot_hashes = gossip_data.value.data.SnapshotHashes;
+
+        var max_inc_hash: ?SlotAndHash = null;
+        for (snapshot_hashes.incremental) |inc_hash| {
+            if (max_inc_hash == null or inc_hash.slot > max_inc_hash.?.slot) {
+                max_inc_hash = inc_hash;
+            }
+        }
+
+        // dont try to download from a slow peer
+        for (blacklist) |black_list_peers| {
+            if (black_list_peers.equals(&peer_contact_info.pubkey)) {
+                result.is_blacklist += 1;
+                continue :search_loop;
+            }
+        }
+
+        valid_peers.appendAssumeCapacity(.{
+            // NOTE: maybe we need to deep clone here due to arraylist sockets?
+            .contact_info = peer_contact_info.*,
+            .full_snapshot = snapshot_hashes.full,
+            .inc_snapshot = max_inc_hash,
+        });
+    }
+    result.is_valid = valid_peers.items.len;
+
+    return result;
+}
+
+/// uses swap remove to remove peers that do not have a hash which is trusted
+pub fn removeUnTrustedSnapshotPeers(
+    allocator: std.mem.Allocator,
+    table: *const GossipTable,
+    trusted_validators: std.ArrayList(Pubkey),
+    available_snapshot_peers: *std.ArrayList(PeerSnapshotHash),
+) !void {
+    var trusted_snapshot_hashes = std.AutoHashMap(
+        SlotAndHash, // full snapshot hash
+        std.AutoHashMap(SlotAndHash, void), // set of incremental snapshots
+    ).init(allocator);
+    defer trusted_snapshot_hashes.deinit();
+
+    // get the hashes of trusted validators
+    var trusted_count: usize = 0;
+    for (trusted_validators.items) |trusted_validator| {
+        const gossip_data = table.get(.{ .SnapshotHashes = trusted_validator }) orelse continue;
+        const trusted_hashes = gossip_data.value.data.SnapshotHashes;
+        trusted_count += 1;
+
+        var max_inc_hash: ?SlotAndHash = null;
+        for (trusted_hashes.incremental) |inc_hash| {
```

Comment:
Why not insert all of these into the set of incremental hashes?

Alternatively, why not abandon the set entirely, and restrict each full snapshot to be associated with only a single max incremental across all trusted validators?

The current behavior is like a medium between the two extremes and I'm curious what the goal is of finding per-validator maxes.

Reply:
incremental snapshots are always associated with a base_slot which corresponds to a full snapshot - so we cant have a single incremental across all validators (different base_slots wouldnt be valid)

it does make sense to include all the incrementals though (which the rust client isnt doing) - eg, with [full_slot, inc_slot] notion: we could have a trusted validator with [19, {19, 12}] and a valid validator download could be [19, 12] 💯 - will make the change

Reply:
4c588e4

Reply:
4c588e4